agent/engine/docker_task_engine.go

```diff
@@ -391,66 +415,116 @@ func (engine *DockerTaskEngine) monitorQueuedTasks(ctx context.Context) {
 		case <-ctx.Done():
 			return
 		case <-engine.monitorQueuedTaskEvent:
-			// Dequeue as many tasks as possible and start wake up their goroutines
+			// Dequeue as many tasks as possible and start waking up their goroutines
+			index := 0
 			for {
-				task, err := engine.topTask()
+				task, err := engine.getTaskByIndex(index)
 				if err != nil {
+					// either the queue is empty or we've traversed through all the tasks in the queue
+					logger.Debug("Unable to get task from the wait queue at the given index", logger.Fields{
+						field.TaskARN: task.Arn,
+						field.Error:   err,
+						"index":       index,
+					})
 					break
 				}
-				dequeuedTask := engine.tryDequeueWaitingTasks(task)
+				dequeuedTask := engine.tryDequeueWaitingTasks(task, index)
 				if !dequeuedTask {
```

Comment:
Q. As index is incrementing through the loop, are these unable-to-dequeue tasks only be validated again when there is a new `engine.monitorQueuedTaskEvent` coming? or this will not happen as we are going to remove them in the follow-up PR? Thanks!

Reply:
tasks in a queue are evaluated in two scenarios,

- when a running task stops
- when there's a new task payload received from ACS. This could either be to _run_ a new task or a _stop_ an existing task (either in the queue or running).